codec-http2/src/main/java/io/netty/handler/codec/http2/PriorityStreamByteAssigner.java

```diff
@@ -0,0 +1,437 @@
+            streamableBytes = newStreamableBytes;
+
+            // Update this branch of the priority tree if the streamable bytes have changed for this node.
+            unassignedStreamableBytesForTreeChanged(delta);
+        }
+
+        void close() {
+            // Unassign all bytes.
+            resetAssigned();
+
+            // Clear the streamable bytes.
+            update(new StreamState() {
+                @Override
+                public Http2Stream stream() {
+                    return stream;
+                }
+
+                @Override
+                public int streamableBytes() {
+                    return 0;
+                }
+
+                @Override
+                public boolean hasFrame() {
+                    return false;
+                }
+            });
+        }
+
+        boolean hasFrame() {
+            return hasFrame;
+        }
+
+        int unassignedStreamableBytes() {
+            return streamableBytes - assigned;
+        }
+
+        int unassignedStreamableBytesForTree() {
+            return unassignedStreamableBytesForTree;
+        }
+    }
+
+    /**
+     * A connection stream visitor that delegates to the user provided visitor.
+     */
+    private class StreamVisitor implements Http2StreamVisitor {
+        boolean visitorDone;
+        Visitor assignmentVisitor;
+        RuntimeException error;
+        boolean inForEachStream;
+
+        void forEachStream(Visitor assignmentVisitor) {
+            assert !inForEachStream;
+            inForEachStream = true;
+            try {
+                this.assignmentVisitor = assignmentVisitor;
+                visitorDone = false;
+                error = null;
+                try {
+                    connection.forEachActiveStream(this);
+                } catch (Http2Exception e) {
+                    // Should never happen since the visitor doesn't throw.
+                    throw new IllegalStateException(e);
+                }
+
+                // If an error was caught when calling back the visitor, throw it now.
+                if (error != null) {
+                    throw error;
+                }
+            } finally {
+                inForEachStream = false;
+            }
+        }
+
+        @Override
+        public boolean visit(Http2Stream stream) {
+            PriorityState state = state(stream);
+            try {
+                if (!visitorDone) {
+                    // Once the user-supplied visitor indicates that it's done, don't call it again.
+                    visitorDone = !assignmentVisitor.visit(stream, state.assigned);
+                }
+            } catch (RuntimeException e) {
+                // Stop calling the visitor, but continue in the loop to reset the assignment for
+                // all remaining states.
+                visitorDone = true;
+                error = e;
+            } finally {
+                state.resetAssigned();
```

Comment:
It looks like this and getting the assignment amount are the primary motivators behind the extra layer of iteration on top of `Http2StreamVisitor`?  Instead of having this extra layer, and the associated state / error checking consider adding 2 methods to `StreamByteAssigner.StreamState`:

```java
interface StreamState {
...
  int assigned();
  void resetAssigned();
}
```

Then just calling these methods from `WRITE_VISITOR` and change its type back to `Http2StreamVisitor`?